aten/src/ATen/native/README.md

```diff
@@ -0,0 +1,109 @@
+ATen native functions are a mechanism to write ATen methods which only
+make use of other ATen operations (e.g., it is not necessary to bind into
+TH/THC code).  These functions are declared in this file and then folded
+into the ATen code generation process.
+
+## Registering it in `native_functions.yaml`
+
+The first step is to write an entry for your function in
+`native_functions.yaml`.  The format is as follows:
+
+```
+- func: func_name(ArgType arg0[=default], ArgType arg1[=default], ...) -> ReturnType
+```
+
+ArgType(s) are allowed to be simple types understood by ATen
+(e.g. `Tensor`, `TensorList`, `IntList`, `int64_t`, `double`, ...).
+`Tensor?` means that the tensor is optional; if it is not passed by
+the user it defaults to an undefined tensor (if you declare an argument
+as optional, you must check if it is the case).
+
+ReturnType is allowed to be any ArgType or tuple combination of ArgTypes(s),
+e.g. `(Tensor, Tensor)` defaults are optional and are only allowed to be numbers
+(e.g. '0' for `int64_t`, '5.0' for `double`)
+
+The ATen code generation process will generate a header corresponding
+to the C++ implementation you will have to write.
+The C++ function declarations won't match the declaration here because
+they will undergo the standard ATen C++ transformations, e.g. use of const-ref
+for non-inplace Tensor arguments (instead of `Tensor` you will
+take `const Tensor&`).  So after you finish writing your header, we recommend
+running a build and getting the declaration from `NativeFunctions.h`
+(find it with `find -name NativeFunctions.h`) to copy paste into your
+C++ file.
+
+The declarations also support the following attributes:
+
+```
+variants: function, method
+```
+
+Controls whether Tensor method (`t.foo()`) or namespace Function (`at::foo()`) is
+generated as a result of this declaration.  If the declaration is a method,
```

Comment:
This isn't true, where and s_where don't have self as the first argument.  The rule is that dispatch happens on self, not that it has to be the first argument.

Reply:
Actually, I'm not even sure what the rule is, because some of these don't have self.  Can you figure out the rule? :).

Reply:
It seems that self is permitted anywhere, and it will just be elided when generating the method variant.